pubpub/pubpub
# feat: activity digest email tool

**Description**

This PR expands on @howebenhowe's work:
- adds the a column `receivesDigestEmail` to the member table
- adds a [`Switch` input](https://github.com/pubpub/pubpub/pull/1712/files#diff-d4d0f63a20184adab7795572538b339f39370fc06a7d52108ae4f1f45fac0bc1R53) to the Activity dashboard page which toggles the member's `receivesDigestEmail` field. It currently only appears for members of a community with either `"admin"` or `"manage"` permission
- added a tool named [`emailActivityDigest`](https://github.com/pubpub/pubpub/pull/1712/files#diff-e1085e0eaeee53bff6ca2b04fc3d6e0b2c796c92439595470ed3c56aea43c0f6) which iterates batches of communities and renders/sends activity digests to each community member where `member.receivesDigestEmail==true`
- widens the `Digest.tsx` template a bit (from 600px to 640px), centers it on the X-axis, and adds an [invisible preview](https://github.com/pubpub/pubpub/pull/1712/files#diff-6395f0fa3f065d18adb2462152cdd22e8d8173aae0e537849da5bf1acc0d8c42R23) that appears in some email clients' subject lines, but not in the email body

**Screenshots**

Gmail subject line with hidden preview:
<img width="934" alt="Screen Shot 2022-01-05 at 3 34 18 PM" src="https://user-images.githubusercontent.com/6402908/148285390-94b1ef1f-3471-4201-89e0-434aba032bc5.png">

Email content (showing increased width and centering):
<img width="533" alt="Screen Shot 2022-01-05 at 3 01 46 PM" src="https://user-images.githubusercontent.com/6402908/148281337-ee8c93d4-1562-4ca6-8ed9-c75b14c68c13.png">

"Enroll" switch:
<img width="1113" alt="Screen Shot 2022-01-05 at 3 46 53 PM" src="https://user-images.githubusercontent.com/6402908/148287610-70eb0b89-d0d8-4655-9712-07f62ea197b0.png">

**Test Steps**

* Give yourself admin or manage permission to a community.
* Visit the Activity dashboard page for that community, and flip the "Enroll in activity digest emails" switch in the top-right of the page switch on (it should be off by default).
* Run the new activity digest tool: `npm run tools emailActivityDigest`
  * _If you want to speed this up, you can copy `where: { subdomain: <community subdomain> }` after line 220 in `server/community/queries.ts`_
* Wait a few seconds and you should get an email!
